credentials/alts/internal/handshaker/handshaker.go

```diff
@@ -334,6 +337,7 @@ func (h *altsHandshaker) processUntilDone(resp *altspb.HandshakerResp, extra []b
 			return resp.Result, extra, nil
 		}
 		buf := make([]byte, frameLimit)
+		h.conn.SetReadDeadline(time.Now().Add(readTimeout))
```

Comment:
If I didn't miss anything, this will affect all future read on this connection (doc: http://go/godoc/net/#Conn).

So no `Read()` will wait for more than 10 seconds, including when the other side is not sending data.
This behavior doesn't sound right to me. We will need to reset this to 0 sometime later (maybe after the first success?)

@cesarghali WDYT?

Reply:
This is a great call. If we want a deadline on this connection, we should set it at the beginning the handshake and reset it to 0 after the handshake is complete, or when it fails. Let's wait on reviewing this PR until we finish debugging the issue that we're currently working on.

Reply:
Why are we using an arbitrary value of `10s` here. We could use the connection timeout set by the user using the following [WithConnectParams](https://pkg.go.dev/google.golang.org/grpc#WithConnectParams) dial option. If the user has not set this dial option, gRPC sets it to the default value specified [here](https://github.com/grpc/grpc/blob/master/doc/connection-backoff.md) which is `20s`.

In either case, the following sequence of events happen:
- gRPC passes the configured connect timeout as a context deadline to the `transport` layer when attempting to create a client transport
- the `transport` package creates a raw conn to the server, and calls into the `credentials` layer with a call to `ClientHandshake()` and passes the same context passed in by gRPC

There are a couple of options here:
- we could have the `transport` layer set the `Deadline` on the rawConn that is passes to the `credentials` layer. That way, `transport` would set the deadline to the one configured by the user, and after the handshake is complete, it would unset the deadline on the rawConn.
- each credentials implementation could use the deadline set on the context passed to it, and use it as they seem fit.

I prefer option #1 for a couple of reasons:
- it makes the deadline available on the rawConn for all credentials implementations
- we do something similar on the server side. gRPC sets the deadline on the rawConn before calling `ServerHandhake()`. On the server side, this is only option available since the `ServerHandshake()` function does not take a context.

@menghanl @dfawley : What do you think?

Reply:
Yes, we should make sure a deadline is set on the conn before we give it to the credentials.  The credentials should not set their own deadline, as it could interfere with the deadline chosen by gRPC (specifically, if it's longer, it could cause undesirable behavior).